### PR TITLE
Breadcrumb collapse

### PR DESCRIPTION
**Pull request by snowystinger**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

### CODE REVIEW

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx [hunk 1]
```diff
@@ -71,43 +91,123 @@ const wrapper = style<BreadcrumbsStyleProps>({
...
+    </Provider>
   );
 }
 
+let BreadcrumbMenu = (props: {items: Array<Node<any>>, onAction: Pick<BreadcrumbsProps<unknown>, 'onAction'>}) => {
+  let {items, onAction} = props;
+  let {direction} = useLocale();
+  let {size, isDisabled} = useContext(InternalBreadcrumbsContext);
+  // TODO localize See more
+  // somehow remove the "link out" icon in Menus?
```
Comment (reidbarber) [inline packages/@react-spectrum/s2/src/Breadcrumbs.tsx:135]: Is this a question for design, or do we definitely want this?

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx [hunk 2]
```diff
@@ -71,43 +91,123 @@ const wrapper = style<BreadcrumbsStyleProps>({
...
+            size
+          }, styles)}>
+          {children}
+        </RACBreadcrumbs>
+      </CollapsingCollection>
+    </Provider>
   );
 }
 
+let BreadcrumbMenu = (props: {items: Array<Node<any>>, onAction: Pick<BreadcrumbsProps<unknown>, 'onAction'>}) => {
```
Comment (reidbarber) [inline packages/@react-spectrum/s2/src/Breadcrumbs.tsx:130]:
```suggestion
let BreadcrumbMenu = (props: {items: Array<Node<any>>, onAction: BreadcrumbsProps<unknown>['onAction']}) => {
```

This should fix the `onAction` types mismatch

**Review by reidbarber — CHANGES_REQUESTED** [review]

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx (hunk 2 above)
Comment (snowystinger) [reply packages/@react-spectrum/s2/src/Breadcrumbs.tsx:130]: ah thank you, i forgot about this syntax

File: packages/@react-spectrum/s2/src/TagGroup.tsx [hunk 3]
```diff
@@ -519,7 +517,7 @@ const avatarSize = {
 
 function Tag({children, textValue, ...props}: TagProps, ref: DOMRef<HTMLDivElement>) {
   textValue ||= typeof children === 'string' ? children : undefined;
-  let ctx = useSlottedContext(TagGroupContext);
+  let ctx = useContext(InternalTagGroupContext);
```
Comment (reidbarber) [inline packages/@react-spectrum/s2/src/TagGroup.tsx:520]: Open for a surprise: https://reactspectrum.blob.core.windows.net/reactspectrum/e1d738d3b452f89444ead0ca392099276b3e9a00/storybook-s2/index.html?path=/story/taggroup--disabled

**Review by reidbarber — CHANGES_REQUESTED** [review]

File: packages/@react-spectrum/s2/src/TagGroup.tsx (hunk 3 above)
Comment (snowystinger) [reply packages/@react-spectrum/s2/src/TagGroup.tsx:520]: hahahaha, that's amazing. also, it startled me

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx [hunk 4]
```diff
@@ -71,43 +91,123 @@ const wrapper = style<BreadcrumbsStyleProps>({
...
-      </BreadcrumbsContext.Provider>
-    </RACBreadcrumbs>
+    <Provider
+      values={[
+        [InternalBreadcrumbsContext, {size, isDisabled}]
+      ]}>
+      <CollapsingCollection containerRef={domRef} onAction={props.onAction}>
+        <RACBreadcrumbs
+          {...otherProps}
+          isDisabled={isDisabled}
```
Comment (reidbarber) [inline packages/@react-spectrum/s2/src/Breadcrumbs.tsx:117]: The first link is still focusable when `isDisabled` is true. Is that correct? This differs from v3.

**Review by reidbarber — COMMENTED** [review]
Looks good, one question

**Review by reidbarber — DISMISSED** [review]

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx [hunk 5]
```diff
@@ -71,43 +91,123 @@ const wrapper = style<BreadcrumbsStyleProps>({
...
+                <Text slot="label">
+                  {item.props.children({size, isInCtx: true, isCurrent: false, isMenu: true})}
+                </Text>
+              </MenuItem>
+            )}
+          </Menu>
+        </MenuTrigger>
+        <ChevronIcon
+          size={size}
+          className={chevronStyles({direction, isMenu: true})} />
```
Comment (LFDanLu) [inline packages/@react-spectrum/s2/src/Breadcrumbs.tsx:156]: Noticed that the chevrons become grayed out when the breadcrumbs are disabled, is that supposed to be the case? Looks like they aren't in the designs

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx [hunk 6]
```diff
@@ -189,47 +291,180 @@ export interface BreadcrumbProps extends Omit<AriaBreadcrumbItemProps, 'children
...
+  useEffect(() => {
+    // Recalculate visible tags when fonts are loaded.
+    document.fonts?.ready.then(() => updateOverflow());
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+  let sliceIndex = collection.size - visibleItems;
+
+  return (
+    <>
+      <HiddenBreadcrumbs items={children} size="M" listRef={listRef} />
```
Comment (LFDanLu) [inline packages/@react-spectrum/s2/src/Breadcrumbs.tsx:455]: I'm was bit surprised that the size provided here can remain as `M` regardless of the actual size set, but it seems like size doesn't actually get used by the breadcrumb styles. Do we still need it then?

**Review by LFDanLu — COMMENTED** [review]
Behavior looks good to me, just a couple of questions

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx (hunk 6 above)
Comment (snowystinger) [reply packages/@react-spectrum/s2/src/Breadcrumbs.tsx:455]: Ah, forgot to do size changes, I'll update

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx (hunk 4 above)
Comment (snowystinger) [reply packages/@react-spectrum/s2/src/Breadcrumbs.tsx:117]: Looks like it was a bug in our Link implementation in RAC. I've updated to fix it.

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx [hunk 7]
```diff
@@ -10,17 +10,36 @@
...
+import {ActionButton as S2ActionButton} from './ActionButton';
 import {size, style} from '../style/spectrum-theme' with { type: 'macro' };
-import {useDOMRef} from '@react-spectrum/utils';
+import {Text} from './Content';
+import {useDOMRef, useResizeObserver} from '@react-spectrum/utils';
+import {useLayoutEffect} from '@react-aria/utils';
 import {useSpectrumContextProps} from './useSpectrumContextProps';
 
+const MIN_VISIBLE_ITEMS = 1;
+const MAX_VISIBLE_ITEMS = 4;
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-spectrum/s2/src/Breadcrumbs.tsx:41]: not intentional, good catch, thanks. found an issue with the 'exact' number of items as max visible too, I've fixed both

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx (hunk 5 above)
Comment (snowystinger) [reply packages/@react-spectrum/s2/src/Breadcrumbs.tsx:156]: looks like you are correct

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx (hunk 1 above)
Comment (snowystinger) [reply packages/@react-spectrum/s2/src/Breadcrumbs.tsx:135]: heard back from design, i need to figure out some way to remove them

Comment (devongovett) [reply packages/@react-spectrum/s2/src/Breadcrumbs.tsx:135]: Hmm really? That would be inconsistent with menu...

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx [hunk 8]
```diff
@@ -189,47 +297,191 @@ export interface BreadcrumbProps extends Omit<AriaBreadcrumbItemProps, 'children
...
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+  let sliceIndex = collection.size - visibleItems;
+
+  return (
+    <>
+      <HiddenBreadcrumbs items={children} size={size} listRef={listRef} />
+      {visibleItems < collection.size ? (
+        <>
+          {children[0].render?.(children[0])}
```
Comment (devongovett) [inline packages/@react-spectrum/s2/src/Breadcrumbs.tsx:475]: Do we want to keep the `showRoot` option from v3?

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx [hunk 9]
```diff
@@ -71,43 +91,123 @@ const wrapper = style<BreadcrumbsStyleProps>({
...
+        <MenuTrigger>
+          <S2ActionButton isDisabled={isDisabled} isQuiet aria-label="See more"><FolderIcon /></S2ActionButton>
+          {/* @ts-ignore - how are these onAction's not compatible? */}
+          <Menu items={items} onAction={onAction}>
+            {(item: Node<any>) => (
+              <MenuItem
+                {...item.props.originalProps}
+                key={item.key}>
+                <Text slot="label">
+                  {item.props.children({size, isInCtx: true, isCurrent: false, isMenu: true})}
```
Comment (devongovett) [inline packages/@react-spectrum/s2/src/Breadcrumbs.tsx:148]: What is `isInCtx`?

File: packages/react-aria-components/src/Breadcrumbs.tsx [hunk 10]
```diff
@@ -31,6 +31,7 @@ export const BreadcrumbsContext = createContext<ContextValue<BreadcrumbsProps<an
 function Breadcrumbs<T extends object>(props: BreadcrumbsProps<T>, ref: ForwardedRef<HTMLOListElement>) {
   [props, ref] = useContextProps(props, ref, BreadcrumbsContext);
   let {CollectionRoot} = useContext(CollectionRendererContext);
+  console.log('Breadcrumbs RAC', props.isDisabled);
```
Comment (LFDanLu) [inline packages/react-aria-components/src/Breadcrumbs.tsx:34]: stray console.log

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx [hunk 11]
```diff
@@ -189,47 +296,190 @@ export interface BreadcrumbProps extends Omit<AriaBreadcrumbItemProps, 'children
...
+  }, []);
+  let sliceIndex = collection.size - visibleItems;
+
+  return (
+    <>
+      <HiddenBreadcrumbs items={children} size={size} listRef={listRef} />
+      {visibleItems < collection.size ? (
+        <>
+          {children[0].render?.(children[0])}
+          <BreadcrumbMenu items={children.slice(1, sliceIndex)} onAction={onAction} />
```
Comment (LFDanLu) [inline packages/@react-spectrum/s2/src/Breadcrumbs.tsx:471]: What if you only have two breadcrumbs? I assume it shouldn't show the breadcrumb menu still? It seems if you collapse it enough when there are only two items that the Breadcrumb menu shows up:

image

**Review by LFDanLu — COMMENTED** [review]
One more thing I noticed

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx (hunk 11 above)
Comment (snowystinger) [reply packages/@react-spectrum/s2/src/Breadcrumbs.tsx:471]: fixed, see storybook from https://github.com/adobe/react-spectrum/pull/6892/commits/ac304c9ace5d2bd4ca6aceae7de360c33d080060 where i temporarily added some new stories to hit the edge cases

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx (hunk 8 above)
Comment (snowystinger) [reply packages/@react-spectrum/s2/src/Breadcrumbs.tsx:470]: following up with spectrum

**Review by LFDanLu — APPROVED** [review]
LGTM, verified the link out arrows aren't there anymore in the stories, breadcrumb text and chevron disabled color and the 2 breadcrumb item case

File: packages/@react-spectrum/s2/src/Breadcrumbs.tsx [hunk 12]
```diff
@@ -71,43 +91,122 @@ const wrapper = style<BreadcrumbsStyleProps>({
...
+    </Provider>
   );
 }
 
+let BreadcrumbMenu = (props: {items: Array<Node<any>>, onAction: BreadcrumbsProps<unknown>['onAction']}) => {
+  let {items, onAction} = props;
+  let {direction} = useLocale();
+  let {size, isDisabled} = useContext(InternalBreadcrumbsContext);
+  // TODO localize See more
+  // somehow remove the "link out" icon in Menus?
```
Comment (reidbarber) [inline packages/@react-spectrum/s2/src/Breadcrumbs.tsx:135]:
```suggestion
```

**Review by reidbarber — DISMISSED** [review]

**Review by LFDanLu — APPROVED** [review]

**Review by ktabors — APPROVED** [review]

**Review by reidbarber — APPROVED** [review]

**Pull request merged**